File: maze.py
import random

def neighbors(cell, size):
    return [[cell[0] + x, cell[1]] for x in [-2, 2] if cell[0] + x > 0 and cell[0] + x < size + 2] + [[cell[0], cell[1] + y] for y in [-2, 2] if cell[1] + y > 0 and cell[1] + y < size + 2]

def generateMaze(size, start=1):
    # randomized prim's algorithm

    wlist = neighbors([start, start], size)
    maze = [['u' for y in range(size + 2)] for x in range(size + 2)]
    maze[start][start] = 'c'
    for wall in wlist: maze[wall[0]][wall[1]] = 's'

    while wlist:
        randWall = random.choice(wlist)
        notEdge = randWall[0] != size + 1 and randWall[1] != size + 1
        nearbyCells = []
        for cell in neighbors(randWall, size):
            value = maze[cell[0]][cell[1]]
            if value == 'c': nearbyCells += [cell]
            elif value == 'u': 
                maze[cell[0]][cell[1]] = 's'
                if notEdge: wlist += [cell]

        chosenCell = random.choice(nearbyCells)
        maze[(randWall[0] + chosenCell[0]) // 2][(randWall[1] + chosenCell[1]) // 2] = 'c'
        if notEdge: maze[randWall[0]][randWall[1]] = 'c'

        wlist.remove(randWall)

    # normalize cell values
    for x in range(1, size + 1):
        if maze[size + 1][x] == 's': maze[size][x] = 'c'
        if maze[x][size + 1] == 's': maze[x][size] = 'c'

    for y, x in enumerate(maze):
        for z in range(len(x)):
            if x[z] != 'c':
                maze[y][z] = 0
            else:
                maze[y][z] = 1

    end = random.choice([i for i in range(size)[1:-1] if maze[-2][i-1] != maze[-2][i]])
    return (maze, [start, start], [size, end])

File: src/js/maze.js
function neighborCells([x, y], size) {
  let neighbors = [];
  for (let value of [-2, 2]) {
    if (x + value > 0 && x + value < size + 2) neighbors.push([x + value, y]);
    if (y + value > 0 && y + value < size + 2) neighbors.push([x, y + value]);
  }
  return neighbors;
}

function generateMaze(size) {
  let start = [1, 1];
  let maze = Array.from({ length: size + 2 }, () =>
    new Array(size + 2).fill("")
  );
  let wallList = neighborCells(start, size);
  for (let wall of wallList) maze[wall[0]][wall[1]] = "s";
  maze[1][1] = "c";

  while (wallList.length) {
    let tentative = wallList[Math.floor(Math.random() * wallList.length)];
    let notEdge = tentative[0] != size + 1 && tentative[1] != size + 1;
    let nearbyCells = neighborCells(tentative, size).filter((cell) => {
      value = maze[cell[0]][cell[1]];
      if (value == "c") return cell;
      if (!value) {
        maze[cell[0]][cell[1]] = "s";
        if (notEdge) wallList.push(cell);
      }
    });

    let chosenCell =
      nearbyCells[Math.floor(Math.random() * nearbyCells.length)];
    maze[(tentative[0] + chosenCell[0]) / 2][
      (tentative[1] + chosenCell[1]) / 2
    ] = "c";
    if (notEdge) maze[tentative[0]][tentative[1]] = "c";

    wallList.splice(wallList.indexOf(tentative), 1);
  }

  // normalize values
  for (let i = 1; i < size + 1; i++) {
    if (maze[size + 1][i] == "s") maze[size][i] = "c";
    if (maze[i][size + 1] == "s") maze[i][size] = "c";
  }

  for (let i = 0; i < size + 2; i++) {
    for (let j = 0; j < size + 2; j++) {
      if (maze[i][j] != "c") maze[i][j] = 0;
      else maze[i][j] = 1;
    }
  }

  let deadEnds = [];
  let verticals = 0;

  for (let i = 1; i < size; i++) {
    if (neighbors([size, i]).reduce((r, d) => r + maze[d[0]][d[1]], 0) == 1) {
      deadEnds.unshift(i);
      verticals++;
    }
    if (neighbors([i, size]).reduce((r, d) => r + maze[d[0]][d[1]], 0) == 1)
      deadEnds.push(i);
  }

  let endIndex = Math.floor(Math.random() * deadEnds.length);

  return [
    maze,
    start,
    endIndex < verticals
      ? [size, deadEnds[endIndex]]
      : [deadEnds[endIndex], size],
  ];
}
